### PR TITLE
HDDS-11013. Ensure version is always set in ContainerCommandRequestProto

### PR DESCRIPTION
**Pull request by swamirishi**
## What changes were proposed in this pull request?
Setting ClientVersion in the ContainerCommandRequestProto object in XceiverGrpcClient send method, to ensure client version is always set on the request object.
## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-11013

## How was this patch tested?
Existing Unit test, no functional change.

### CODE REVIEW

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/ClientVersion.java [hunk 1]
```diff
@@ -28,7 +28,7 @@
 /**
  * Versioning for protocol clients.
  */
-public enum ClientVersion implements ComponentVersion {
+public enum ClientVersion implements ComponentVersion, Comparable<ClientVersion> {
```
Comment (hemantk-12) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/ClientVersion.java:31]: Is it intentional change?

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/storage/ContainerProtocolCalls.java [hunk 2]
```diff
@@ -97,6 +98,28 @@ public final class ContainerProtocolCalls  {
...
 
+  /**
+   * Creates a ContainerCommandRequestProto with version set.
+   */
+  public static ContainerCommandRequestProto.Builder getContainerCommandRequestProtoBuilder(int version) {
+    return getContainerCommandRequestProtoBuilder(null, version);
+  }
+
+  public static ContainerCommandRequestProto.Builder getContainerCommandRequestProtoBuilder() {
+    return getContainerCommandRequestProtoBuilder(ClientVersion.CURRENT.toProtoValue());
```
Comment (hemantk-12) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/storage/ContainerProtocolCalls.java:109]: nit:
```suggestion
    return getContainerCommandRequestProtoBuilder(null, ClientVersion.CURRENT.toProtoValue());
```

**Review by hemantk-12 — COMMENTED** [review]
Thanks @swamirishi for the patch. In general, change looks good to me. Left a comment about one change in ClientVersion.

nit: I don't think it is a good idea to provide a helper function on top of builder. I guess you have done it this way to make the version mandatory field and `ContainerCommandRequestProto.newBuilder()` is generated by protobuf.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/ClientVersion.java (hunk 1 above)
Comment (swamirishi) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/ClientVersion.java:31]: No thanks for pointing out.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/storage/ContainerProtocolCalls.java (hunk 2 above)
Comment (swamirishi) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/storage/ContainerProtocolCalls.java]: done

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/ClientVersion.java (hunk 1 above)
Comment (swamirishi) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/ClientVersion.java]: removed it.

**Comment by swamirishi** [comment]
> Thanks @swamirishi for the patch. In general, change looks good to me. Left a comment about one change in ClientVersion.
> 
> nit: I don't think it is a good idea to provide a helper function on top of builder. I guess you have done it this way to make the version mandatory field and `ContainerCommandRequestProto.newBuilder()` is generated by protobuf.

Yeah I wanted to mandate client versions are being set to the latest value on all the request. I don't think there is a programatic way of doing this in protobuf. From what I know we can only set a constant value in proto file.

**Comment by swamirishi** [comment]
We can deprecate this method call, when we would stop supporting older clients on the server side.

**Review by hemantk-12 — APPROVED** [review]
LGTM+1.

**Review by adoroszlai — COMMENTED** [review]
Thanks @swamirishi for extracting this.

There are two leftover calls to `newBuilder` in `ContainerProtocolCalls` (`echo` and `readContainerFromAllNodes`) and two in `TestHddsDispatcher` (`newPutSmallFile` and `getWriteChunkRequest0`).

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/storage/ContainerProtocolCalls.java [hunk 3]
```diff
@@ -97,6 +98,28 @@ public final class ContainerProtocolCalls  {
...
+
+  public static ContainerCommandRequestProto.Builder getContainerCommandRequestProtoBuilder() {
+    return getContainerCommandRequestProtoBuilder(null, ClientVersion.CURRENT.toProtoValue());
+  }
+
+  public static ContainerCommandRequestProto.Builder getContainerCommandRequestProtoBuilder(
+      ContainerCommandRequestProto req, int version) {
+    return (req == null ?
+        ContainerCommandRequestProto.newBuilder() : ContainerCommandRequestProto.newBuilder(req)).setVersion(version);
+  }
```
Comment (adoroszlai) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/storage/ContainerProtocolCalls.java:116]: `req` may already have `version` set.  `newBuilder(req)` preserves that in the builder, but `setVersion` overwrites it.  This is a problem if called via `getContainerCommandRequestProtoBuilder(req)`, without explicit version.

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientGrpc.java [hunk 4]
```diff
@@ -337,8 +338,7 @@ private XceiverClientReply sendCommandWithTraceIDAndRetry(
 
     return TracingUtil.executeInNewSpan(spanName,
         () -> {
-          ContainerCommandRequestProto finalPayload =
-              ContainerCommandRequestProto.newBuilder(request)
+          ContainerCommandRequestProto finalPayload = getContainerCommandRequestProtoBuilder(request)
                   .setTraceID(TracingUtil.exportCurrentSpan()).build();
```
Comment (adoroszlai) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientGrpc.java:342]: Would this work?

```suggestion
          ContainerCommandRequestProto.Builder builder = getContainerCommandRequestProtoBuilder(request)
              .setTraceID(TracingUtil.exportCurrentSpan());
          if (!request.hasVersion()) {
            builder.setVersion(ClientVersion.CURRENT.toProtoValue());
          }
          return sendCommandWithRetry(builder.build(), validators);
```

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientGrpc.java [hunk 5]
```diff
@@ -490,8 +490,7 @@ public XceiverClientReply sendCommandAsync(
 
     try (Scope ignored = GlobalTracer.get().activateSpan(span)) {
 
-      ContainerCommandRequestProto finalPayload =
-          ContainerCommandRequestProto.newBuilder(request)
+      ContainerCommandRequestProto finalPayload = getContainerCommandRequestProtoBuilder(request)
```
Comment (adoroszlai) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientGrpc.java:493]: Also set version conditionally here.

**Review by adoroszlai — CHANGES_REQUESTED** [review]
Thanks again @swamirishi for the patch.

One question: instead of replacing all the calls, how about setting `version` in `sendCommand(Async)`?  That would simplify this change a whole lot.  It would also avoid potential problems caused by accidentally using the plain builder instead of the one created by the central functions.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/storage/ContainerProtocolCalls.java (hunk 3 above)
Comment (swamirishi) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/storage/ContainerProtocolCalls.java:116]: If that is the case we should just send a builder than sending the request object. That is also one viable change.

Comment (swamirishi) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/storage/ContainerProtocolCalls.java:116]: or we can just add a validation check and throw an exception if the version is not set.

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientGrpc.java (hunk 4 above)
Comment (swamirishi) [reply hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientGrpc.java:343]: done

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientGrpc.java [hunk 6]
```diff
@@ -277,9 +278,14 @@ public ContainerCommandResponseProto sendCommand(
     List<DatanodeDetails> datanodeList = pipeline.getNodes();
     HashMap<DatanodeDetails, CompletableFuture<ContainerCommandResponseProto>>
             futureHashMap = new HashMap<>();
+    ContainerCommandRequestProto.Builder builder = ContainerCommandRequestProto.newBuilder(request);
```
Comment (hemantk-12) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientGrpc.java:281]: nit: create a helper function something like`getVersionRequest()` and reuse it in other places.

**Review by hemantk-12 — APPROVED** [review]
LGTM.

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientGrpc.java [hunk 7]
```diff
@@ -277,9 +278,14 @@ public ContainerCommandResponseProto sendCommand(
...
+    if (!request.hasVersion()) {
+      builder.setVersion(ClientVersion.CURRENT.toProtoValue());
+    }
+    ContainerCommandRequestProto finalRequest = builder.build();
```
Comment (adoroszlai) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientGrpc.java:285]: Here we only conditionally need to create `newBuilder()` and then `build()`.

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientGrpc.java [hunk 8]
```diff
@@ -337,10 +343,13 @@ private XceiverClientReply sendCommandWithTraceIDAndRetry(
 
     return TracingUtil.executeInNewSpan(spanName,
         () -> {
-          ContainerCommandRequestProto finalPayload =
+          ContainerCommandRequestProto.Builder finalPayload =
```
Comment (adoroszlai) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientGrpc.java:346]: nit: `finalPayload` can be renamed to `builder`.

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientGrpc.java [hunk 9]
```diff
@@ -490,12 +499,14 @@ public XceiverClientReply sendCommandAsync(
 
     try (Scope ignored = GlobalTracer.get().activateSpan(span)) {
 
-      ContainerCommandRequestProto finalPayload =
+      ContainerCommandRequestProto.Builder finalPayload =
```
Comment (adoroszlai) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientGrpc.java:502]: nit: `finalPayload` can be renamed to `builder`.

File: hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestBlockOutputStreamCorrectness.java [hunk 10]
```diff
@@ -231,6 +232,11 @@ public Pipeline getPipeline() {
     public XceiverClientReply sendCommandAsync(
         ContainerCommandRequestProto request
     ) {
+      final ContainerCommandRequestProto.Builder finalRequest = ContainerCommandRequestProto.newBuilder(request);
```
Comment (adoroszlai) [inline hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestBlockOutputStreamCorrectness.java:235]: nit: `finalRequest` can be renamed to `builder`.

File: hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestBlockOutputStreamCorrectness.java [hunk 11]
```diff
@@ -231,6 +232,11 @@ public Pipeline getPipeline() {
...
+      if (!request.hasVersion()) {
+        finalRequest.setVersion(ClientVersion.CURRENT.toProtoValue());
+      }
+      request = finalRequest.build();
```
Comment (adoroszlai) [inline hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestBlockOutputStreamCorrectness.java:239]: Here we only conditionally need to create `newBuilder()` and then `build()`.

File: hadoop-hdds/common/src/test/java/org/apache/hadoop/ozone/container/ContainerTestHelper.java [hunk 12]
```diff
@@ -549,16 +555,15 @@ public static byte[] generateData(int length, boolean random) {
    * @return
    */
   public static ContainerCommandRequestProto getDummyCommandRequestProto(
-      ContainerProtos.Type cmdType) {
-    final Builder builder =
-        ContainerCommandRequestProto.newBuilder()
+      ClientVersion clientVersion, ContainerProtos.Type cmdType, int replicaIndex) {
+    final Builder builder = ContainerCommandRequestProto.newBuilder().setVersion(clientVersion.toProtoValue())
```
Comment (adoroszlai) [inline hadoop-hdds/common/src/test/java/org/apache/hadoop/ozone/container/ContainerTestHelper.java:559]: nit: 

```suggestion
    final Builder builder = ContainerCommandRequestProto.newBuilder()
        .setVersion(clientVersion.toProtoValue())
```

**Review by adoroszlai — COMMENTED** [review]
Thanks a lot @swamirishi for updating the patch.  I think it's much simpler now.  Only few nits picked.

**Review by adoroszlai — APPROVED** [review]
Thanks @swamirishi for updating the patch, LGTM.

**Comment by swamirishi** [comment]
Thank your for the review @hemantk-12 @adoroszlai

**Pull request merged**